src/main/java/org/mastodon/graph/branch/BranchGraph.java

```diff
@@ -0,0 +1,230 @@
+ * <li><code>bv1</code> links to <code>v3</code>,
+ * <li>all linked vertices in-between linked to the branch edge between
+ * <code>bv0</code> and <code>bv1</code>,
+ * <li>as well as all the linked edges in the branch.
+ * <li>The branch edge between <code>bv0</code> and <code>bv1</code> links to
+ * the outgoing edge of <code>v0</code>.
+ * </ul>
+ * <p>
+ * For instance, a graph laid as following:
+ *
+ * <pre>
+ *                 v3 &#8594; v4 &#8594; v5
+ *              /
+ * v0 &#8594; v1 &#8594; v2
+ *              \
+ *                 v6 &#8594; v7 &#8594; v8
+ * </pre>
+ *
+ * will be represented by the following branch graph:
+ *
+ * <pre>
+ *            bv2
+ *          /
+ * bv0 &#8594; bv1
+ *          \
+ *            bv3
+ * </pre>
+ *
+ * In the example above, <code>v0</code>, <code>v2</code>, <code>v5</code> and
+ * <code>v8</code> are <b>branch extremities</b>; they are linked to a unique
+ * branch vertex in the branch graph. <code>v1</code>, <code>v3</code>,
+ * <code>v4</code>, <code>v6</code> and <code>v7</code> belongs to a branch.
+ * They link to a branch edge.
+ *
+ * <p>
+ * The branch graph is based on a non-simple directed graph. There might be more
+ * than one edge between the same source and target branch vertices. This
+ * happens for instance when there is a diamond-like shape in the linked graph:
+ *
+ * <pre>
+ *     v1 &#8594; v2 &#8594; v3
+ *   /             \
+ * v0               v7
+ *   \             /
+ *     v4 &#8594; v5 &#8594; v6
+ * </pre>
+ *
+ * In the branch graph, this becomes:
+ *
+ * <pre>
+ *     ____
+ *    /    \
+ * bv0      bv1
+ *    \____/
+ * </pre>
+ *
+ * The branch graph can also handle ring-link structures in the linked graph. In
+ * that case, such a ring;
+ *
+ * <pre>
+ *     v1 &#8594; v2 &#8594; v3
+ *   /             \
+ * v0               v4
+ *   \             /
+ *     v7 &#8592; v6 &#8592; v5
+ * </pre>
+ *
+ * is represented by a single vertex having a loop-edge:
+ *
+ * <pre>
+ *    _
+ * bv0 \
+ *  \__/
+ * </pre>
+ *
+ * In such a loop, the vertex linked to <code>bv0</code> and the edge linked to
+ * the branch edge are determined by the order in which the vertices and edges
+ * are added in the linked graph.
+ * <p>
+ * The branch graph is defined and tested only for <b>simple, directed
+ * graphs</b> as linked graph. Using any other classes of graphs will result in
+ * unexpected behavior.
+ *
+ * @param <BV>
+ *            the type of the branch vertices.
+ * @param <BE>
+ *            the type of the branch edges.
+ * @param <V>
+ *            the type of linked vertices.
+ * @param <E>
+ *            the type of linked edges.
+ */
+public interface BranchGraph<
+	BV extends Vertex< BE >,
+	BE extends Edge< BV >,
+	V extends Vertex< E >,
+	E extends Edge< V > >
+		extends ListenableReadOnlyGraph< BV, BE >
+{
+
+	/**
+	 * Returns the edge linked to the specified branch edge. The linked edge is
+	 * the single outgoing edge of the branch extremity.
```

Comment:
It is not necessarily true, that a branch extremity has a single outgoing edge. (?) Can you make this more clear?